File: code/mathematical_algorithms/src/std/std.c
#include <stdio.h>
#include <math.h>


// calculate the mean of a given array having n elements
double
__mean(double arr[], size_t n)
{
    double sum = 0;
    for (size_t i = 0; i < n; i++) {
        sum = sum + arr[i];
    }
    double mean = sum / n;
    return (mean);
}

// calculate the standard deviation of a given array having n elements
double 
std(double arr[], size_t n)
{
    double mean = __mean(arr, n);
    double std = 0;
    for (size_t i = 0; i < n; i++) {
        std = std + (arr[i] - mean) * (arr[i] - mean);
    }
    return (sqrt(std / n));
}

int 
main()
{
    double arr[] = { 12, 65, 91, 52, 18, 72 };
    printf("%lf", std(arr, sizeof(arr) / sizeof(arr[0]))); // 28.41

    return (0);
}

File: code/mathematical_algorithms/src/std/std.js
function sd(arr) {
  // calculate standard deviation of array of numbers
  var arr_sum = 0;
  var arr_square_sum = 0;
  for (var i = 0; i < arr.length; ++i) {
    arr_sum += arr[i]; // sum of values in array
    arr_square_sum += arr[i] * arr[i]; // sum of squares of values in array
  }

  var arr_mean = arr_sum / arr.length; // this is E[X]
  var arr_square_mean = arr_square_sum / arr.length; // this is E[X^2]
  var variance = arr_square_mean - arr_mean * arr_mean; // Var(X) = E[X^2] - E[X]^2

  return Math.sqrt(variance);
}

// tests:
sd([12, 65, 91, 52, 18, 72]); // returns 28.41, as expected
sd([-10.6]); // returns 0, as expected
sd([-1, -1, 0, 1, 1]); // returns 0.8944, as expected
